database/skylink.go

```diff
@@ -143,34 +123,130 @@ func (db *DB) AddServerForSkylink(ctx context.Context, skylink string, server st
+//     "$or": [
+//         { "lock_expires" : { "$exists": false }},
+//         { "lock_expires" : { "$lt": new Date() }}
+//     ]
+// })
+func (db *DB) FindAndLockUnderpinned(ctx context.Context, server string, minPinners int) (skymodules.Skylink, error) {
+	// First try to fetch a skylink which is locked by the current server. This
+	// is our mechanism for proactively recovering from files being left locked
+	// after a server crash.
+	sl, err := db.findLockedSkylink(ctx, server)
+	if err == nil {
+		return sl, nil
+	}
+	// No files locked by the current server were found, look for unlocked ones.
+	filter := bson.M{
+		"unpin": false,
+		// Pinned by fewer than the minimum number of servers.
+		"$expr": bson.M{"$lt": bson.A{bson.M{"$size": "$servers"}, minPinners}},
+		// Not pinned by the given server.
+		"servers": bson.M{"$nin": bson.A{server}},
+		// Unlocked.
+		"$or": bson.A{
+			bson.M{"lock_expires": bson.M{"$exists": false}},
+			bson.M{"lock_expires": bson.M{"$lt": time.Now().UTC().Truncate(time.Millisecond)}},
+		},
+	}
+	update := bson.M{
+		"$set": bson.M{
+			"locked_by":    server,
+			"lock_expires": time.Now().UTC().Add(LockDuration).Truncate(time.Millisecond),
+		},
+	}
+	opts := options.FindOneAndUpdate().SetReturnDocument(options.After)
+	sr := db.staticDB.Collection(collSkylinks).FindOneAndUpdate(ctx, filter, update, opts)
+	if sr.Err() == mongo.ErrNoDocuments {
+		return skymodules.Skylink{}, ErrSkylinkNoExist
+	}
+	if sr.Err() != nil {
+		return skymodules.Skylink{}, sr.Err()
+	}
+	var result struct {
+		Skylink string
+	}
+	err = sr.Decode(&result)
+	if err != nil {
+		return skymodules.Skylink{}, errors.AddContext(err, "failed to decode result")
+	}
+	return SkylinkFromString(result.Skylink)
+}
+
+// findLockedSkylink fetches a skylink that's locked by the current server, if
+// one exists.
+func (db *DB) findLockedSkylink(ctx context.Context, server string) (skymodules.Skylink, error) {
+	filter := bson.M{
+		"locked_by":    server,
+		"lock_expires": bson.M{"$gt": time.Now().UTC().Truncate(time.Millisecond)},
+		"unpin":        false,
+	}
+	sr := db.staticDB.Collection(collSkylinks).FindOne(ctx, filter)
+	if sr.Err() != nil {
+		return skymodules.Skylink{}, sr.Err()
+	}
+	var result struct {
+		Skylink string
+	}
+	err := sr.Decode(&result)
+	if err != nil {
+		return skymodules.Skylink{}, err
+	}
+	return SkylinkFromString(result.Skylink)
+}
+
+// UnlockSkylink removes the lock on the skylink put while we're trying to pin
+// it to a new server.
+func (db *DB) UnlockSkylink(ctx context.Context, skylink skymodules.Skylink, server string) error {
+	filter := bson.M{
+		"skylink":   skylink.String(),
+		"locked_by": server,
+	}
+	update := bson.M{
+		"$set": bson.M{
+			"locked_by":    "",
+			"lock_expires": time.Time{},
+		},
+	}
+	ur, err := db.staticDB.Collection(collSkylinks).UpdateOne(ctx, filter, update)
+	if errors.Contains(err, mongo.ErrNoDocuments) || ur.ModifiedCount == 0 {
```

Comment:
Ha! Good thing I've written a test for this, otherwise I would have never remembered why it's there!

If we remove that check and we try to unlock a skylink from a server that hasn't locked it, the method will return with `nil` error, indicating success, even when no unlocking has actually happened. That's why we want to make sure that we managed to find the skylink *and* we managed to modify its record before we return that `nil` error.

Reply:
ah got it.

This case seems similar to another comment I made. Feels like a new error `ErrNoSkylinksLocked` might be useful to better handle this. 

Or to change the return to `(unlocked bool, err error)` so that the caller can decide if the condition is problematic. 

Because if there are no skylnks being locked by the server, that is ok, so returning an error doesn't feel needed. The caller just needs to know if they unlocked anything. 

I'm fine handling this as a follow up, since it doesn't change the core functionality, just adjusts the code readability IMO.

Reply:
Done.